actioncable/test/channel/stream_test.rb

```diff
@@ -144,6 +144,116 @@ class StreamTest < ActionCable::TestCase
+
+        assert_equal 0, subscribers_of(connection).size
+
+        channel.stream_from "room_one"
+        channel.stream_from "room_two"
+
+        wait_for_async
+        assert_equal 2, subscribers_of(connection).size
+
+        channel2 = ChatChannel.new connection, "User B"
+        channel2.subscribe_to_channel
+
+        channel2.stream_from "room_one"
+        wait_for_async
+
+        subscribers = subscribers_of(connection)
+
+        assert_equal 2, subscribers.size
+        assert_equal 2, subscribers["room_one"].size
+        assert_equal 1, subscribers["room_two"].size
+
+        channel.stop_all_streams
+
+        subscribers = subscribers_of(connection)
+        assert_equal 1, subscribers_of(connection).size
+        assert_equal 1, subscribers["room_one"].size
+      end
+    end
+
+    test "stop_stream_from" do
+      run_in_eventmachine do
+        connection = TestConnection.new
+
+        channel = ChatChannel.new connection, "User A"
+        channel.subscribe_to_channel
+
+        channel.stream_from "room_one"
+        channel.stream_from "room_two"
+
+        channel2 = ChatChannel.new connection, "User B"
+        channel2.subscribe_to_channel
+
+        channel2.stream_from "room_one"
+
+        subscribers = subscribers_of(connection)
+
+        wait_for_async
+
+        assert_equal 2, subscribers.size
+        assert_equal 2, subscribers["room_one"].size
+        assert_equal 1, subscribers["room_two"].size
+
+        channel.stop_stream_from "room_one"
+
+        subscribers = subscribers_of(connection)
+
+        assert_equal 2, subscribers.size
+        assert_equal 1, subscribers["room_one"].size
+        assert_equal 1, subscribers["room_two"].size
+      end
+    end
+
+    test "stop_stream_for" do
+      run_in_eventmachine do
+        connection = TestConnection.new
+
+        channel = ChatChannel.new connection, "User A"
+        channel.subscribe_to_channel
+
+        channel.stream_for Room.new(1)
+        channel.stream_for Room.new(2)
+
+        channel2 = ChatChannel.new connection, "User B"
+        channel2.subscribe_to_channel
+
+        channel2.stream_for Room.new(1)
+
+        subscribers = subscribers_of(connection)
+
+        wait_for_async
+
+        assert_equal 2, subscribers.size
+
+        assert_equal 2, subscribers[ChatChannel.broadcasting_for(Room.new(1))].size
+        assert_equal 1, subscribers[ChatChannel.broadcasting_for(Room.new(2))].size
+
+        channel.stop_stream_for Room.new(1)
+
+        subscribers = subscribers_of(connection)
+
+        assert_equal 2, subscribers.size
+        assert_equal 1, subscribers[ChatChannel.broadcasting_for(Room.new(1))].size
+        assert_equal 1, subscribers[ChatChannel.broadcasting_for(Room.new(2))].size
+      end
+    end
+
+    private
+      def subscribers_of(connection)
+        connection
+          .pubsub
+          .subscriber_map
+          .instance_variable_get(:@subscribers)
+      end
```

Comment:
Testing against internal implementation details is brittle, but feels appropriate in this case. Alternatively, we'd need to broadcast and check which subscribers received it.

Reply:
Agree, maybe it's possible to switch test pubsub adapter from `actioncable/test/stubs/test_adapter.rb` to `actioncable/lib/action_cable/subscription_adapter/inline.rb`